src/wkt/src/message.rs

```diff
@@ -39,19 +39,17 @@ pub trait Message {
     where
         Self: serde::de::DeserializeOwned,
     {
-        let map: Map = map
-            .iter()
-            .filter_map(|(k, v)| {
-                if k == "@type" {
-                    return None;
-                }
-                Some((k.clone(), v.clone()))
-            })
-            .collect();
-        serde_json::from_value::<Self>(serde_json::Value::Object(map)).map_err(Error::deser)
+        from_object(map)
     }
 }
 
+/// Write the serialization of `T` flatly into a map.
+///
+/// This is used for types where the JSON serialization of `T` is an object.
+///
+/// `T` is a message with fields that cannot be named `@type`, so flattening the
+/// object saves some bytes over the wire (as compared with encoding the
+/// serialization into a `value` field).
```

Comment:
I am not sure that is the motivation. We are just trying to comply with the ProtoJSON spec:

https://protobuf.dev/programming-guides/json/

Reply:
I think I was trying to explain why the proto spec is the way it is. I will just defer to the spec.